serde_with/src/guide/serde_as_transformations.md

```diff
@@ -388,6 +389,39 @@ value: u32,
 "value": "666",
 ```
 
+## Struct fields as map keys
+
+[`KeyValueMap`]
+
+This conversion is possible for structs and maps, using the `$key$` field.
+Tuples, tuple structs, and sequences are supported by turning the first value into the map key.
+
+Each of the `SimpleStruct`s
+
+```ignore
+// Somewhere there is a collection:
+// #[serde_as(as = "KeyValueMap<_>")]
+// Vec<SimpleStruct>,
+
+#[derive(Serialize, Deserialize)]
+struct SimpleStruct {
+    b: bool,
+    // The field named `$key$` will become the map key
+    #[serde(rename = "$key$")]
```

Comment:
This is really difficult with serde, since serde has no way to extend its data model or any sidechannel information. Thus changing the name or using newtypes is an accepted way of changing behavior. Using a special name was actually taken from https://docs.rs/serde-xml-rs/latest/serde_xml_rs/index.html.

I am not sure how a `#[serde(map_key)]` would work. The general problem is, that this would require shipping a custom derive macro. I am kinda sympathetic to replacing `serde_derive` with something better, this feels way too much for this use case.

Deserializing from two different formats should work with `alias`. Different serialization forms are not really possible with serde anyway, since there is only one `Serialize` implementation possible.

One thing might be a nice change. Instead of using a hardcoded `$key$` value, it could instead take an argument how the field is called anyway. Something like
```rust
#[serde_as(as = r#"KeyValueMap<"id", _>"#)]
Vec<SimpleStruct>,
```
By using this const generic argument, changes to the original struct can be minimized. This does require the unstable `adt_const_params` feature (https://github.com/rust-lang/rust/issues/95174). But I will not that down as a future improvement.

Reply:
Fair enough